### PR TITLE
add peinjector

### PR DESCRIPTION
**Pull request by thealpiste**
This extension allows to inject a given shellcode into a portable executable, it is based in the following project [https://github.com/JonDoNym/peinjector](https://github.com/JonDoNym/peinjector). A post module was added in the client side to manage de injection.

EDIT: your link was broken and driving me nuts - @wwebb-r7

### CODE REVIEW

**Comment by wwebb-r7** [comment]
Hi,

At a glance this looks pretty strong for a first-time contribution, good job.  With regards to that, don't be discouraged if it takes awhile to eventually get merged.  Payload PR's progress a bit slower than the framework side of things.

**Comment by thealpiste** [comment]
thank you, i will wait for your feedback

**Comment by sempervictus** [comment]
Looks rather neat at first pass, definitely interesting. Will try to get it tested in the near term.

**Comment by wwebb-r7** [comment]
@thealpiste - I started reviewing some of the code, and may add to it over the next few days.  If something I say is not clear, let me know.  I've never used the github review feature before and I'm liable to screw some quotation up.

**Comment by thealpiste** [comment]
Changes were updated, it is failing because i think there is something outdated in the java payload. Should i fix it?

**Comment by timwr** [comment]
Apologies @thealpiste the test fail is unrelated to your changes!
https://github.com/rapid7/metasploit-payloads/pull/243

**Comment by wwebb-r7** [comment]
@thealpiste sorry about that.  I've been on vacation for awhile and playing catch up this week.  It'll probably be Monday before I give this another go, but I'll get to it.

**Comment by thealpiste** [comment]
@wwebb-r7 just for curiosity, are you still working on this?

**Comment by wwebb-r7** [comment]
@thealpiste I actually set aside today to give it another go.  I had come across some weird build issues on my end during the last effort that needed to be worked out first, then the holidays hit. I'll try to give you an update later today on the status of where I'm at.

**Comment by wwebb-r7** [comment]
```
1>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppBuild.targets(1190,5): warning MSB8030: The linker switch "Minimum Required Version" requires "SubSystem" to be set.  Without "SubSystem", the "Minimum Required Version" would not be passed to linker and could prevent to the output binary from running on older Operating Systems.
1>LINK : fatal error LNK1181: cannot open input file ',5.02'
========== Build: 0 succeeded, 1 failed, 4 up-to-date, 0 skipped ==========
```

on the 64 bit build.  Setting the subsystem under `project properties -> Linker -> System` will clear both up.

Additionally, after testing and discussing things with others, I feel like this would be more friendly as a true Meterpreter extension, as in one that can be used within a session, without having to background out to msfconsole.  Before I go infecting files on a target, I'd like to be able to run some basic stuff like `ls` and such, as well as having the ability to run one command as opposed to loading up a post module and setting all the options.  If you're up to modifying it, great.  If not, I can see about doing myself sometime in the future.

**Comment by thealpiste** [comment]
Regarding the first error i cannot reproduce it on my local machine, i have **"Windows (/SUBSYSTEM:WINDOWS)"** option selected. I will check it again before next commit.

Regarding to the meterpreter extension, at first i created both, the post module and the meterpreter extension, but then i decided to continue only with the post module, i would suggest to have both.

Some examples would be appreciated, i was thinking in something like this:

-h => Show help
-t => Remote target file to be infected
-p => Selected payload
-lhost
-lport
-o =>aditional options

**Comment by wwebb-r7** [comment]
Let me double check on the linker settings, I wasn't the only one who experienced it.  We test by merging into master and building from that, so something may be getting lost in translation.

As far as options, those sound fine to me.  I'll drag up some documentation that hopefully isn't too out of date on building new extensions.  For other examples, there's the driver extension PR I have open.  Some of it is currently a mess but the TLV definitions and handlers should be good.

**Comment by thealpiste** [comment]
In the first commit i uploaded a meterpreter extension by mistake, it was working fine, i will move all the logic from the post module to the meterpreter extension.

**Comment by thealpiste** [comment]
Meterpreter extension interface added.

**Comment by wwebb-r7** [comment]
Nice.  I likely won't get a chance to go over it until tonight or tomorrow morning but it's on my TODO list

**Comment by bwatters-r7** [comment]
Also, if you submit a fix, could you standardize line endings?  Some are in unix style, some are windows.

**Comment by thealpiste** [comment]
The  issues  were  fixed,  i  will  be  reviewing  the  metasploit  module

File: c/meterpreter/source/extensions/peinjector/libpeinfect.c [hunk 1]
```diff
@@ -0,0 +1,819 @@
...
+
+	/* Optional Header present? */
+	if (!((out->optional_header_32.Magic == NT_OPTIONAL_32_MAGIC)
+		|| (out->optional_header_64.Magic == NT_OPTIONAL_64_MAGIC))) {
+		return false;
+	}
+
+	/* Clean Header fields (ASLR, Checksum, Signature DataDir, ...) */
+	if (in->remove_integrity) {
+		__peinfect_clean_header(in, out);
```
Comment (bwatters-r7) [inline c/meterpreter/source/extensions/peinjector/libpeinfect.c]: This is likely the culprit....

**Comment by bwatters-r7** [comment]
@thealpiste I was having some trouble with those blank commits, so I did a force push on top of them so they disappear and re-implemented them.  We could coordinate on it to try and salvage them, but the changes were super simple, and this was just easy and it lets me land it now.

**Comment by bwatters-r7** [comment]
# Release Notes
This adds the ability to insert arbitrary shellcode into a PE file as a thread.

**Pull request merged**